[Bug] Idea ultimate 2022.2.2 启动报异常

java.lang.ExceptionInInitializerError
	at com.itangcent.common.spi.SpiUtils.loadService(SpiUtils.kt:53)
	at com.itangcent.common.spi.SpiUtils$loadService$1.apply(SpiUtils.kt:31)
	at java.base/java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1708)
	at com.itangcent.common.spi.SpiUtils.loadService(SpiUtils.kt:30)
	at com.itangcent.common.kit.KitUtilsKt.toJson(KitUtils.kt:33)
	at com.itangcent.idea.plugin.configurable.AbstractEasyApiConfigurable.isModified(AbstractEasyApiConfigurable.kt:38)
	at com.intellij.openapi.options.ex.ConfigurableWrapper.isModified(ConfigurableWrapper.java:174)
	at com.intellij.openapi.options.newEditor.SettingsEditor.checkModifiedInternal(SettingsEditor.java:525)
	at com.intellij.openapi.options.newEditor.SettingsEditor.checkModifiedForItem(SettingsEditor.java:519)
	at com.intellij.openapi.options.newEditor.SettingsEditor.checkModified(SettingsEditor.java:507)
	at com.intellij.openapi.options.newEditor.SettingsEditor$5.updateCurrent(SettingsEditor.java:234)
	at com.intellij.openapi.options.newEditor.ConfigurableEditor.lambda$select$2(ConfigurableEditor.java:252)
	at com.intellij.openapi.util.ExecutionCallback.setExecuted(ExecutionCallback.java:44)
	at com.intellij.openapi.util.ActionCallback.setDone(ActionCallback.java:72)
	at com.intellij.ui.CardLayoutPanel.select(CardLayoutPanel.java:124)
	at com.intellij.ui.CardLayoutPanel.lambda$selectLater$0(CardLayoutPanel.java:135)
	at com.intellij.openapi.application.TransactionGuardImpl.runWithWritingAllowed(TransactionGuardImpl.java:209)
	at com.intellij.openapi.application.TransactionGuardImpl.access$100(TransactionGuardImpl.java:21)
	at com.intellij.openapi.application.TransactionGuardImpl$1.run(TransactionGuardImpl.java:191)
	at com.intellij.openapi.application.impl.ApplicationImpl.runIntendedWriteActionOnCurrentThread(ApplicationImpl.java:881)
	at com.intellij.openapi.application.impl.ApplicationImpl$3.run(ApplicationImpl.java:513)
	at com.intellij.openapi.application.impl.FlushQueue.doRun(FlushQueue.java:75)
	at com.intellij.openapi.application.impl.FlushQueue.runNextEvent(FlushQueue.java:118)
	at com.intellij.openapi.application.impl.FlushQueue.flushNow(FlushQueue.java:42)
	at java.desktop/java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:318)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:779)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:730)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:724)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:399)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:86)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:749)
	at com.intellij.ide.IdeEventQueue.defaultDispatchEvent(IdeEventQueue.java:918)
	at com.intellij.ide.IdeEventQueue._dispatchEvent(IdeEventQueue.java:766)
	at com.intellij.ide.IdeEventQueue.lambda$dispatchEvent$6(IdeEventQueue.java:450)
	at com.intellij.openapi.progress.impl.CoreProgressManager.computePrioritized(CoreProgressManager.java:791)
	at com.intellij.ide.IdeEventQueue.lambda$dispatchEvent$7(IdeEventQueue.java:449)
	at com.intellij.openapi.application.TransactionGuardImpl.performActivity(TransactionGuardImpl.java:113)
	at com.intellij.ide.IdeEventQueue.performActivity(IdeEventQueue.java:624)
	at com.intellij.ide.IdeEventQueue.lambda$dispatchEvent$8(IdeEventQueue.java:447)
	at com.intellij.openapi.application.impl.ApplicationImpl.runIntendedWriteActionOnCurrentThread(ApplicationImpl.java:881)
	at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.java:493)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:207)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:128)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:121)
	at java.desktop/java.awt.WaitDispatchSupport$2.run(WaitDispatchSupport.java:191)
	at java.desktop/java.awt.WaitDispatchSupport$4.run(WaitDispatchSupport.java:236)
	at java.desktop/java.awt.WaitDispatchSupport$4.run(WaitDispatchSupport.java:234)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:318)
	at java.desktop/java.awt.WaitDispatchSupport.enter(WaitDispatchSupport.java:234)
	at java.desktop/java.awt.Dialog.show(Dialog.java:1080)
	at com.intellij.openapi.ui.impl.DialogWrapperPeerImpl$MyDialog.show(DialogWrapperPeerImpl.java:712)
	at com.intellij.openapi.ui.impl.DialogWrapperPeerImpl.show(DialogWrapperPeerImpl.java:453)
	at com.intellij.openapi.ui.DialogWrapper.doShow(DialogWrapper.java:1672)
	at com.intellij.openapi.ui.DialogWrapper.show(DialogWrapper.java:1630)
	at com.intellij.ide.actions.ShowSettingsUtilImpl.showSettingsDialog(ShowSettingsUtilImpl.java:90)
	at com.intellij.openapi.wm.impl.welcomeScreen.CustomizeTab$buildComponent$1$5$1.invoke(CustomizeTabFactory.kt:206)
	at com.intellij.openapi.wm.impl.welcomeScreen.CustomizeTab$buildComponent$1$5$1.invoke(CustomizeTabFactory.kt:65)
	at com.intellij.ui.components.ActionLink$2.actionPerformed(ActionLink.kt:36)
	at java.desktop/javax.swing.AbstractButton.fireActionPerformed(AbstractButton.java:1972)
	at java.desktop/javax.swing.AbstractButton$Handler.actionPerformed(AbstractButton.java:2313)
	at java.desktop/javax.swing.DefaultButtonModel.fireActionPerformed(DefaultButtonModel.java:405)
	at java.desktop/javax.swing.DefaultButtonModel.setPressed(DefaultButtonModel.java:262)
	at java.desktop/javax.swing.plaf.basic.BasicButtonListener.mouseReleased(BasicButtonListener.java:279)
	at java.desktop/java.awt.Component.processMouseEvent(Component.java:6648)
	at java.desktop/javax.swing.JComponent.processMouseEvent(JComponent.java:3392)
	at java.desktop/java.awt.Component.processEvent(Component.java:6413)
	at java.desktop/java.awt.Container.processEvent(Container.java:2266)
	at java.desktop/java.awt.Component.dispatchEventImpl(Component.java:5022)
	at java.desktop/java.awt.Container.dispatchEventImpl(Container.java:2324)
	at java.desktop/java.awt.Component.dispatchEvent(Component.java:4854)
	at java.desktop/java.awt.LightweightDispatcher.retargetMouseEvent(Container.java:4948)
	at java.desktop/java.awt.LightweightDispatcher.processMouseEvent(Container.java:4575)
	at java.desktop/java.awt.LightweightDispatcher.dispatchEvent(Container.java:4516)
	at java.desktop/java.awt.Container.dispatchEventImpl(Container.java:2310)
	at java.desktop/java.awt.Window.dispatchEventImpl(Window.java:2802)
	at java.desktop/java.awt.Component.dispatchEvent(Component.java:4854)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:781)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:730)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:724)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:399)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:86)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:97)
	at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:754)
	at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:752)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:399)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:86)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:751)
	at com.intellij.ide.IdeEventQueue.defaultDispatchEvent(IdeEventQueue.java:918)
	at com.intellij.ide.IdeEventQueue.dispatchMouseEvent(IdeEventQueue.java:840)
	at com.intellij.ide.IdeEventQueue._dispatchEvent(IdeEventQueue.java:763)
	at com.intellij.ide.IdeEventQueue.lambda$dispatchEvent$6(IdeEventQueue.java:450)
	at com.intellij.openapi.progress.impl.CoreProgressManager.computePrioritized(CoreProgressManager.java:791)
	at com.intellij.ide.IdeEventQueue.lambda$dispatchEvent$7(IdeEventQueue.java:449)
	at com.intellij.openapi.application.TransactionGuardImpl.performActivity(TransactionGuardImpl.java:113)
	at com.intellij.ide.IdeEventQueue.performActivity(IdeEventQueue.java:624)
	at com.intellij.ide.IdeEventQueue.lambda$dispatchEvent$8(IdeEventQueue.java:447)
	at com.intellij.openapi.application.impl.ApplicationImpl.runIntendedWriteActionOnCurrentThread(ApplicationImpl.java:881)
	at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.java:493)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:207)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:128)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:117)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:105)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:92)
Caused by: java.lang.IllegalStateException: Recursive update
	at java.base/java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1763)
	at com.itangcent.common.spi.SpiUtils.loadService(SpiUtils.kt:30)
	at com.itangcent.common.spi.SpiUtilsKt.<clinit>(SpiUtils.kt:95)
	... 104 more

我没办法重现这个问题。

> Caused by: java.lang.IllegalStateException: Recursive update
at java.base/java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1763)
at com.itangcent.common.spi.SpiUtils.loadService(SpiUtils.kt:30)
at com.itangcent.common.spi.SpiUtilsKt.(SpiUtils.kt:95)
... 104 more

可以看一下这里的more?

---

Anyway, 我先尝试处理SpiUtils.loadService潜在的Recursive问题。

--- a/README.md
+++ b/README.md
@@ -32,26 +32,12 @@
 * [配置](https://easyyapi.com/setting/index.html)
 * [demo](https://easyyapi.com/demo/index.html)
 
-## 贡献
+## Contributing
 
-***您可以通过以下途径帮助easy-yapi变得更好***
+You can propose a feature request opening an issue or a pull request.
 
-* 讨论 [新的功能](https://github.com/tangcent/easy-yapi/issues?q=label%3Aenhancement)
-  or提交 [BUG](https://github.com/tangcent/easy-yapi/issues?q=label%3Abug): [Issues](https://github.com/tangcent/easy-yapi/issues)
+Here is a list of contributors:
 
-* 开发 [新的功能](https://github.com/tangcent/easy-yapi/pulls?q=+label%3Aenhancement)
-  or修复 [BUG](https://github.com/tangcent/easy-yapi/pulls?q=label%3Abug): [PR](https://github.com/tangcent/easy-yapi/pulls)
-
-* 完善测试 [codecov](https://codecov.io/gh/tangcent/easy-yapi)
-
-* 帮助维护文档: [easyyapi.com](https://github.com/easyyapi/easyyapi.github.io/tree/dev)
-  请将您的PR提交到dev分支
-
-*
-维护推荐配置: [.recommend.easy.api.config](https://github.com/tangcent/easy-yapi/blob/master/idea-plugin/src/main/resources/.recommend.easy.api.config)
-
-* 丰富支持的三方配置: [third](https://github.com/tangcent/easy-yapi/tree/master/third)
-
-* 丰富demo: [spring-demo](https://github.com/Earth-1610/spring-demo) | [web-demo](https://github.com/Earth-1610/web-demo)
-
-* 修复IDE兼容问题: [intellij-kotlin](https://github.com/Earth-1610/intellij-kotlin/pulls?q=label%3Acompatibility+)
+<a href="https://github.com/tangcent/easy-yapi/graphs/contributors">
+  <img src="https://contrib.rocks/image?repo=tangcent/easy-yapi" />
+</a>

--- a/README_en.md
+++ b/README_en.md
@@ -14,13 +14,13 @@
 - [demo](https://github.com/Earth-1610/web-demo)
 
 # Table of Contents
-
-* 1 [Feature](#Feature)
-* 2 Doc
-  * 2.1 [Javadoc](#Javadoc)
-  * 2.2 [KDoc](#KDoc)
-* 3 [Installation](#Installation)
-* 4 [Guide](#Guide)
+<!-- TOC -->
+* [Feature](#feature)
+* [Javadoc](#javadoc)
+* [KDoc](#kdoc)
+* [Installation](#installation)
+* [Contributing](#contributing)
+<!-- TOC -->
 
 ## Feature
 
@@ -60,72 +60,12 @@ Installation
 restart **IDE**.
 
 
-## Guide
+## Contributing
 
-* ExportApi(0.8.2.1+)
-```textCode
-    1. Open existed Spring Controller File Or Select files or directories from project navigation
-    You can use by this : "alt shift E(windows)/ctrl E(mac)"
-    2. Select apis and channel
-    3. Click [✔️] button or press enter key
-```
+You can propose a feature request opening an issue or a pull request.
 
-* ExportYapi
-```textCode
-    There are two ways to export api.
-    1. Open existed Spring Controller File
-    You can use by this : "Right click in the file -> generate... " or use its shortcuts "[Alt + Insert]/[Ctrl+Enter]" , then
-    choose the action "ExportYapi"
-    2. Select files or directories from project navigation
-    You can use by this : "Click [Code -> ExportYapi] in top"
-```
+Here is a list of contributors:
 
-* ExportPostman
-```textCode
-    There are two ways to export api.
-    1. Open existed Spring Controller File
-    You can use by this : "Right click in the file -> generate... " or use its shortcuts "[Alt + Insert]/[Ctrl+Enter]" , then
-    choose the action "ExportPostman"
-    2. Select files or directories from project navigation
-    You can use by this : "Click [Code -> ExportPostman] in top"
-```
-
-* How to export to postman automatically?
-
-```text
-    Click [Preference -> Other Setting -> EasyApi]
-    set postman privateToken
-    If you do not have a privateToken of postman,
-    you can easily generate one by heading over to the Postman Integrations Dashboard
-    [https://go.postman.co/integrations/services/pm_pro_api]
-```
-
-* Quick API requests from code
-
-```textCode
-    Open existed Spring Controller File
-    You can use by this : "Right click in the file -> generate... " or use its shortcuts "[Alt + Insert]/[Ctrl+Enter]" , then
-    choose the action "Call"
-```
-
-* ApiDashBoard
-```textCode
-    It is easily to export api in current project to postman by dragging
-    You can use by this : "Click [Code -> ApiDashBoard] in top"
-```
-
-* YApiDashBoard
-```textCode
-    It is easily to export api in current project to yapi by dragging
-    You can use by this : "Click [Code -> YApiDashBoard] in top"
-```
-
-* ExportMarkdown(Beta)
-```textCode
-    There are two ways to export api.
-    1. Open existed Spring Controller File
-    You can use by this : "Right click in the file -> generate... " or use its shortcuts "[Alt + Insert]/[Ctrl+Enter]" , then
-    choose the action "ExportMarkdown"
-    2. Select files or directories from project navigation
-    You can use by this : "Click [Code -> ExportMarkdown] in top"
-```
+<a href="https://github.com/tangcent/easy-yapi/graphs/contributors">
+  <img src="https://contrib.rocks/image?repo=tangcent/easy-yapi" />
+</a>

--- a/common-api/build.gradle
+++ b/common-api/build.gradle
@@ -26,7 +26,7 @@ dependencies {
 
     implementation "org.jetbrains.kotlin:kotlin-reflect:$kotlin_version"
 
-    implementation('com.itangcent:commons:1.3.8') {
+    implementation('com.itangcent:commons:1.3.9') {
         exclude group: 'com.google.inject'
         exclude group: 'com.google.code.gson'
     }

--- a/idea-plugin/build.gradle
+++ b/idea-plugin/build.gradle
@@ -78,17 +78,17 @@ dependencies {
 //    implementation fileTree(dir: 'libs', include: ['*.jar'])
 
 
-    implementation('com.itangcent:intellij-idea:1.3.8') {
+    implementation('com.itangcent:intellij-idea:1.3.9') {
         exclude group: 'com.google.inject'
         exclude group: 'com.google.code.gson'
     }
 
-    implementation('com.itangcent:intellij-kotlin-support:1.3.8') {
+    implementation('com.itangcent:intellij-kotlin-support:1.3.9') {
         exclude group: 'com.google.inject'
         exclude group: 'com.google.code.gson'
     }
 
-//    implementation('com.itangcent:intellij-scala-support:1.3.8') {
+//    implementation('com.itangcent:intellij-scala-support:1.3.9') {
 //        exclude group: 'com.google.inject'
 //        exclude group: 'com.google.code.gson'
 //    }
@@ -122,7 +122,7 @@ dependencies {
     // https://mvnrepository.com/artifact/org.mockito/mockito-inline
     testImplementation group: 'org.mockito', name: 'mockito-inline', version: '3.11.0'
 
-    testImplementation('com.itangcent:intellij-idea-test:1.3.8') {
+    testImplementation('com.itangcent:intellij-idea-test:1.3.9') {
         exclude group: 'com.nhaarman.mockitokotlin2', module: 'mockito-kotlin'
     }
 }

--- a/idea-plugin/src/main/kotlin/com/itangcent/idea/config/CachedResourceResolver.kt
+++ b/idea-plugin/src/main/kotlin/com/itangcent/idea/config/CachedResourceResolver.kt
@@ -3,6 +3,7 @@ package com.itangcent.idea.config
 import com.google.inject.Inject
 import com.google.inject.Singleton
 import com.google.inject.name.Named
+import com.itangcent.common.logger.Log
 import com.itangcent.common.logger.traceError
 import com.itangcent.common.utils.TimeSpanUtils
 import com.itangcent.idea.plugin.settings.helper.HttpSettingsHelper
@@ -24,6 +25,9 @@ import java.util.concurrent.TimeUnit
 
 @Singleton
 open class CachedResourceResolver : DefaultResourceResolver() {
+    companion object : Log() {
+        private const val URL_CACHE_EXPIRE = "url.cache.expire"
+    }
 
     @Inject(optional = true)
     @Named("projectCacheRepository")
@@ -121,6 +125,3 @@ open class CachedResourceResolver : DefaultResourceResolver() {
         private fun timeOut() = httpSettingsHelper.httpTimeOut(TimeUnit.MILLISECONDS)
     }
 }
-
-private const val URL_CACHE_EXPIRE = "url.cache.expire"
-private val LOG = com.intellij.openapi.diagnostic.Logger.getInstance(CachedResourceResolver::class.java)

--- a/idea-plugin/src/main/kotlin/com/itangcent/idea/plugin/api/ClassApiExporterHelper.kt
+++ b/idea-plugin/src/main/kotlin/com/itangcent/idea/plugin/api/ClassApiExporterHelper.kt
@@ -3,6 +3,7 @@ package com.itangcent.idea.plugin.api
 import com.google.inject.Inject
 import com.google.inject.Singleton
 import com.intellij.psi.*
+import com.itangcent.common.logger.Log
 import com.itangcent.common.model.Doc
 import com.itangcent.common.utils.KV
 import com.itangcent.common.utils.notNullOrBlank
@@ -62,6 +63,9 @@ open class ClassApiExporterHelper {
 
     @Inject
     protected lateinit var messagesHelper: MessagesHelper
+
+    companion object : Log()
+
     fun extractParamComment(psiMethod: PsiMethod): KV<String, Any>? {
         val subTagMap = docHelper!!.getSubTagMapOfDocComment(psiMethod, "param")
         if (subTagMap.isEmpty()) {
@@ -237,5 +241,3 @@ open class ClassApiExporterHelper {
         }
     }
 }
-
-private val LOG = com.intellij.openapi.diagnostic.Logger.getInstance(ClassApiExporterHelper::class.java)

--- a/idea-plugin/src/main/kotlin/com/itangcent/idea/plugin/api/DefaultMethodInferHelper.kt
+++ b/idea-plugin/src/main/kotlin/com/itangcent/idea/plugin/api/DefaultMethodInferHelper.kt
@@ -5,6 +5,7 @@ import com.google.inject.Singleton
 import com.intellij.psi.*
 import com.intellij.psi.util.*
 import com.itangcent.common.kit.KVUtils
+import com.itangcent.common.logger.Log
 import com.itangcent.common.logger.traceError
 import com.itangcent.common.utils.*
 import com.itangcent.idea.plugin.settings.SettingBinder
@@ -141,6 +142,7 @@ class DefaultMethodInferHelper : MethodInferHelper {
                 copy.sortByDescending { pointOf(it) }
                 return copy
             }
+
             is Map<*, *> -> {
                 if (this.isEmpty() || this.size == 1) {
                     return this
@@ -156,6 +158,7 @@ class DefaultMethodInferHelper : MethodInferHelper {
                 }
                 return copy
             }
+
             is Variable -> return this.getValue().cleanInvalidKeys()
         }
         return this
@@ -168,9 +171,11 @@ class DefaultMethodInferHelper : MethodInferHelper {
             is Collection<*> -> {
                 return 2 + obj.map { pointOf(it) }.sum()
             }
+
             is Map<*, *> -> {
                 return 3 + obj.entries.map { pointOf(it.key) + pointOf(it.value) }.sum()
             }
+
             is Variable -> return pointOf(obj.getValue())
             is String -> return if (obj.isEmpty()) 1 else 2
         }
@@ -420,6 +425,7 @@ class DefaultMethodInferHelper : MethodInferHelper {
                 val fixArgs = Array<Any?>(method.parameterCount) { null }
                 method.invoke(caller, fixArgs)
             }
+
             args == null || args.isEmpty() -> method.invoke(caller)
             args.size == 1 -> method.invoke(caller, args[0])
             args.size == 2 -> method.invoke(caller, args[0], args[1])
@@ -514,7 +520,7 @@ class DefaultMethodInferHelper : MethodInferHelper {
     private var collection_add_method: PsiMethod? = null
     private var collection_addAll_method: PsiMethod? = null
 
-    companion object {
+    companion object : Log() {
 
         private const val ALLOW_QUICK_CALL: Int = 0b0001
 
@@ -563,20 +569,23 @@ class DefaultMethodInferHelper : MethodInferHelper {
                         obj.entries.forEach { copy[valueOf(it.key)] = valueOf(it.value) }
                         return copy
                     }
+
                     obj is Array<*> -> {
                         val copy = LinkedList<Any?>()
                         obj.any { copy.add(valueOf(it)) }
                         return copy.toArray()
                     }
+
                     obj is Collection<*> -> {
                         val copy = LinkedList<Any?>()
                         obj.any { copy.add(valueOf(it)) }
                         return copy
                     }
+
                     else -> obj
                 }
             } catch (e: Exception) {
-                LOG.error("failed compute valueOf $obj", e)
+                LOG.traceError("failed compute valueOf $obj", e)
                 return null
             }
         }
@@ -592,12 +601,15 @@ class DefaultMethodInferHelper : MethodInferHelper {
                 is MutableMap<*, *> -> {
                     obj.entries.any { needCompute(it.key, deep + 1) || needCompute(it.value, deep + 1) }
                 }
+
                 is Array<*> -> {
                     obj.any { needCompute(it, deep + 1) }
                 }
+
                 is Collection<*> -> {
                     obj.any { needCompute(it, deep + 1) }
                 }
+
                 else -> false
             }
         }
@@ -717,16 +729,20 @@ class DefaultMethodInferHelper : MethodInferHelper {
             psiType is PsiArrayType -> {
                 return psiClassHelper.getTypeObject(psiType, context, simpleJsonOption)
             }
+
             jvmClassHelper!!.isCollection(psiType) -> {   //list type
                 return psiClassHelper.getTypeObject(psiType, context, jsonOption)
             }
+
             jvmClassHelper.isMap(psiType) -> {   //map type
                 return psiClassHelper.getTypeObject(psiType, context, simpleJsonOption)
             }
+
             jvmClassHelper.isEnum(psiType) -> {
                 //return "" by default
                 return ""
             }
+
             else -> {
                 val typeCanonicalText = psiType.canonicalText
                 if (typeCanonicalText.contains('<') && typeCanonicalText.endsWith('>')) {
@@ -754,12 +770,15 @@ class DefaultMethodInferHelper : MethodInferHelper {
             psiType is PsiArrayType -> {
                 return psiClassHelper.getTypeObject(psiType, context, simpleJsonOption)
             }
+
             jvmClassHelper!!.isCollection(psiType) -> {   //list type
                 return psiClassHelper.getTypeObject(psiType, context, simpleJsonOption)
             }
+
             jvmClassHelper.isMap(psiType) -> {   //map type
                 return psiClassHelper.getTypeObject(psiType, context, simpleJsonOption)
             }
+
             else -> {
                 val paramCls = PsiUtil.resolveClassInClassTypeOnly(psiType) ?: return null
                 if (ruleComputer!!.computer(ClassRuleKeys.TYPE_IS_FILE, paramCls) == true) {
@@ -902,28 +921,34 @@ class DefaultMethodInferHelper : MethodInferHelper {
                     }
                     return variable
                 }
+
                 is PsiIfStatement -> {
                     statement.elseBranch?.let { processStatement(it) }
                     statement.thenBranch?.let { processStatement(it) }
                 }
+
                 is PsiBlockStatement -> {
                     processBlock(statement.codeBlock)
                 }
+
                 is PsiTryStatement -> {
                     statement.tryBlock?.let { processBlock(it) }
                     for (catchBlock in statement.catchBlocks) {
                         processBlock(catchBlock)
                     }
                 }
+
                 is PsiForStatement -> {
                     statement.initialization?.let { processStatement(it) }
                     statement.update?.let { processStatement(it) }
                     statement.body?.let { processStatement(it) }
                 }
+
                 is PsiForeachStatement -> {
                     statement.iteratedValue?.let { processExpression(it) }
                     statement.body?.let { processStatement(it) }
                 }
+
                 is PsiWhileStatement -> statement.body?.let { processStatement(it) }
                 is PsiDoWhileStatement -> statement.body?.let { processStatement(it) }
 
@@ -933,15 +958,18 @@ class DefaultMethodInferHelper : MethodInferHelper {
                         processElement(declaredElement)
                     }
                 }
+
                 is PsiReturnStatement -> {
                     val returnValue = statement.returnValue
                     if (returnValue != null) {
                         returnVal = findComplexResult(returnVal, processExpression(returnValue))
                     }
                 }
+
                 is PsiThrowStatement -> {
                     //ignore
                 }
+
                 else -> {
                     methodReturnInferHelper.logger.debug("no matched statement:${statement::class} - ${statement.text}")
                 }
@@ -1002,9 +1030,11 @@ class DefaultMethodInferHelper : MethodInferHelper {
                     }
                     return variable
                 }
+
                 is PsiEnumConstant -> {
                     return resolveEnumFields(psiElement)
                 }
+
                 is PsiField -> {
                     if (methodReturnInferHelper.jvmClassHelper!!.isStaticFinal(psiElement)) {
                         return processStaticField(psiElement)
@@ -1020,32 +1050,40 @@ class DefaultMethodInferHelper : MethodInferHelper {
                         DirectVariable { methodReturnInferHelper.getSimpleFields(psiElement.type, psiElement) }
                     }
                 }
+
                 is PsiParameter -> {
                     return psiElement.name?.let { findVariable(it) }
                 }
+
                 is PsiAnonymousClass -> {
                     processAnonymousClass(psiElement)
                     return null
                 }
+
                 is PsiClassInitializer -> {
                     processBlock(psiElement.body)
                 }
+
                 is PsiKeyword -> {
                     //todo:any keyword return null??
                     return null
                 }
+
                 is PsiReferenceParameterList -> {
                     //todo:what does PsiReferenceParameterList mean
                     return null
                 }
+
                 is PsiWhiteSpace -> {
                     //ignore white space
                     return null
                 }
+
                 is PsiJavaCodeReferenceElement -> {
                     //todo:what does PsiJavaCodeReferenceElement mean
                     return null
                 }
+
                 is PsiExpressionList -> {
                     val list = ArrayList<Any?>()
                     for (expression in psiElement.expressions) {
@@ -1053,6 +1091,7 @@ class DefaultMethodInferHelper : MethodInferHelper {
                     }
                     return list
                 }
+
                 is PsiArrayAccessExpression -> {
                     val array = processExpression(psiElement.arrayExpression) ?: return null
                     if (array is Array<*> && array.size > 0) {
@@ -1064,9 +1103,11 @@ class DefaultMethodInferHelper : MethodInferHelper {
                     }
 
                 }
+
                 is PsiLambdaExpression -> {
                     return psiElement.body?.let { processElement(it) }
                 }
+
                 else -> {
                     //ignore
 //                    methodReturnInferHelper.logger.debug("no matched ele ${psiElement::class.qualifiedName}:${psiElement.text}")
@@ -1096,6 +1137,7 @@ class DefaultMethodInferHelper : MethodInferHelper {
 
                     return le
                 }
+
                 is PsiNewExpression -> return processNewExpression(psiExpression)
                 is PsiCallExpression -> {
 
@@ -1120,6 +1162,7 @@ class DefaultMethodInferHelper : MethodInferHelper {
                     }
                     return ret
                 }
+
                 is PsiReferenceExpression -> {
 
                     val qualifierExpression = psiExpression.qualifierExpression
@@ -1135,6 +1178,7 @@ class DefaultMethodInferHelper : MethodInferHelper {
                     val resolve = psiExpression.resolve()
                     if (resolve != null) return processElement(resolve, qualifier)
                 }
+
                 is PsiThisExpression -> return getThis()
                 is PsiLiteralExpression -> return psiExpression.value
                 is PsiBinaryExpression -> return processBinaryExpression(psiExpression)
@@ -1180,6 +1224,7 @@ class DefaultMethodInferHelper : MethodInferHelper {
                         }
                     }
                 }
+
                 JavaTokenType.MINUS -> {
                     val lOperand = valueOf(processExpression(psiExpression.lOperand)) ?: return null
                     val rOperand = psiExpression.rOperand?.let { valueOf(processExpression(it)) }
@@ -1195,6 +1240,7 @@ class DefaultMethodInferHelper : MethodInferHelper {
                         }
                     }
                 }
+
                 JavaTokenType.ASTERISK -> {
                     val lOperand = valueOf(processExpression(psiExpression.lOperand)) ?: return null
                     val rOperand = psiExpression.rOperand?.let { valueOf(processExpression(it)) }
@@ -1210,6 +1256,7 @@ class DefaultMethodInferHelper : MethodInferHelper {
                         }
                     }
                 }
+
                 JavaTokenType.DIV -> {
                     val lOperand = valueOf(processExpression(psiExpression.lOperand)) ?: return null
                     val rOperand = psiExpression.rOperand?.let { valueOf(processExpression(it)) }
@@ -1225,6 +1272,7 @@ class DefaultMethodInferHelper : MethodInferHelper {
                         }
                     }
                 }
+
                 JavaTokenType.AND -> {
                     val lOperand = valueOf(processExpression(psiExpression.lOperand)) ?: return null
                     val rOperand = psiExpression.rOperand?.let { valueOf(processExpression(it)) }
@@ -1237,6 +1285,7 @@ class DefaultMethodInferHelper : MethodInferHelper {
                         }
                     }
                 }
+
                 JavaTokenType.OR -> {
                     val lOperand = valueOf(processExpression(psiExpression.lOperand)) ?: return null
                     val rOperand = psiExpression.rOperand?.let { valueOf(processExpression(it)) }
@@ -1249,6 +1298,7 @@ class DefaultMethodInferHelper : MethodInferHelper {
                         }
                     }
                 }
+
                 JavaTokenType.NE, JavaTokenType.LE, JavaTokenType.LE, JavaTokenType.LT, JavaTokenType.GE, JavaTokenType.GT -> return true
                 else -> {
                     return valueOf(processExpression(psiExpression.lOperand))
@@ -1273,6 +1323,7 @@ class DefaultMethodInferHelper : MethodInferHelper {
                         }
                     }
                 }
+
                 JavaTokenType.EQEQ, JavaTokenType.NE -> {
                     return true
                 }
@@ -1498,27 +1549,32 @@ class DefaultMethodInferHelper : MethodInferHelper {
                     statement.elseBranch?.let { processStatement(it) }
                     statement.thenBranch?.let { processStatement(it) }
                 }
+
                 is PsiBlockStatement -> {
                     processBlock(statement.codeBlock)
                 }
+
                 is PsiTryStatement -> {
                     statement.tryBlock?.let { processBlock(it) }
                     for (catchBlock in statement.catchBlocks) {
                         processBlock(catchBlock)
                     }
                 }
+
                 is PsiExpressionStatement -> processExpression(statement.expression)
                 is PsiDeclarationStatement -> {
                     for (declaredElement in statement.declaredElements) {
                         processElement(declaredElement)
                     }
                 }
+
                 is PsiReturnStatement -> {
                     val returnValue = statement.returnValue
                     if (returnValue != null) {
                         returnVal = findComplexResult(returnVal, processExpression(returnValue))
                     }
                 }
+
                 else -> {
                     throw IllegalArgumentException("Quickly Infer Failed")
                 }
@@ -1538,13 +1594,16 @@ class DefaultMethodInferHelper : MethodInferHelper {
                     }
                     throw IllegalArgumentException("Quickly Infer Failed")
                 }
+
                 is PsiAnonymousClass -> {
                     processAnonymousClass(psiElement)
                     return null
                 }
+
                 is PsiClassInitializer -> {
                     return processBlock(psiElement.body)
                 }
+
                 else -> {
                     throw IllegalArgumentException("Quickly Infer Failed")
                 }
@@ -1563,6 +1622,7 @@ class DefaultMethodInferHelper : MethodInferHelper {
                         throw IllegalArgumentException("Quickly Infer Failed")
                     }
                 }
+
                 is PsiReferenceExpression -> {
                     val resolve = psiExpression.resolve()
                     if (resolve != null) return processElement(resolve)
@@ -1572,6 +1632,7 @@ class DefaultMethodInferHelper : MethodInferHelper {
 //                    }
                     throw IllegalArgumentException("Quickly Infer Failed")
                 }
+
                 is PsiLiteralExpression -> return psiExpression.value
                 else -> {
                     throw IllegalArgumentException("Quickly Infer Failed")
@@ -1648,5 +1709,3 @@ class DefaultMethodInferHelper : MethodInferHelper {
 }
 
 private typealias LazyAction = () -> Unit
-
-private val LOG = com.intellij.openapi.diagnostic.Logger.getInstance(DefaultMethodInferHelper::class.java)

--- a/idea-plugin/src/main/kotlin/com/itangcent/idea/plugin/api/export/postman/DefaultPostmanApiHelper.kt
+++ b/idea-plugin/src/main/kotlin/com/itangcent/idea/plugin/api/export/postman/DefaultPostmanApiHelper.kt
@@ -3,6 +3,7 @@ package com.itangcent.idea.plugin.api.export.postman
 import com.google.gson.JsonObject
 import com.google.gson.internal.LazilyParsedNumber
 import com.google.inject.Inject
+import com.itangcent.common.logger.Log
 import com.itangcent.common.logger.traceError
 import com.itangcent.common.utils.KV
 import com.itangcent.common.utils.asInt
@@ -431,7 +432,7 @@ open class DefaultPostmanApiHelper : PostmanApiHelper {
         }
     }
 
-    companion object {
+    companion object : Log() {
         const val POSTMAN_HOST = "https://api.getpostman.com"
 
         //const val IMPOREDAPI = "$POSTMANHOST/import/exported"
@@ -449,5 +450,3 @@ open class DefaultPostmanApiHelper : PostmanApiHelper {
     }
 
 }
-
-private val LOG = com.intellij.openapi.diagnostic.Logger.getInstance(DefaultPostmanApiHelper::class.java)

--- a/idea-plugin/src/main/kotlin/com/itangcent/idea/plugin/api/export/postman/PostmanApiExporter.kt
+++ b/idea-plugin/src/main/kotlin/com/itangcent/idea/plugin/api/export/postman/PostmanApiExporter.kt
@@ -3,6 +3,7 @@ package com.itangcent.idea.plugin.api.export.postman
 import com.google.inject.Inject
 import com.google.inject.Singleton
 import com.itangcent.cache.CacheSwitcher
+import com.itangcent.common.logger.Log
 import com.itangcent.common.logger.traceError
 import com.itangcent.common.model.Request
 import com.itangcent.common.utils.*
@@ -41,6 +42,8 @@ class PostmanApiExporter {
 
     @Inject
     private val classApiExporterHelper: ClassApiExporterHelper? = null
+
+    companion object : Log()
 
     fun export() {
         try {
@@ -175,5 +178,3 @@ class PostmanApiExporter {
         postmanApiHelper.updateCollection(collectionId, collectionInfo)
     }
 }
-
-private val LOG = com.intellij.openapi.diagnostic.Logger.getInstance(PostmanApiExporter::class.java)

--- a/idea-plugin/src/main/kotlin/com/itangcent/idea/plugin/api/export/postman/PostmanCachedApiHelper.kt
+++ b/idea-plugin/src/main/kotlin/com/itangcent/idea/plugin/api/export/postman/PostmanCachedApiHelper.kt
@@ -3,6 +3,7 @@ package com.itangcent.idea.plugin.api.export.postman
 import com.google.inject.Inject
 import com.google.inject.Singleton
 import com.itangcent.cache.CacheSwitcher
+import com.itangcent.common.logger.Log
 import com.itangcent.common.utils.notNullOrEmpty
 import com.itangcent.idea.binder.DbBeanBinderFactory
 import com.itangcent.idea.plugin.settings.helper.PostmanSettingsHelper
@@ -337,9 +338,7 @@ class PostmanCachedApiHelper : DefaultPostmanApiHelper(), CacheSwitcher {
         var workspaceDetail: PostmanWorkspace? = null
     )
 
-    companion object {
+    companion object : Log() {
         val NULL_POSTMAN_INFO_CACHE = PostmanInfoCache()
     }
 }
-
-private val LOG = com.intellij.openapi.diagnostic.Logger.getInstance(PostmanCachedApiHelper::class.java)

--- a/idea-plugin/src/main/kotlin/com/itangcent/idea/plugin/configurable/AbstractEasyApiConfigurable.kt
+++ b/idea-plugin/src/main/kotlin/com/itangcent/idea/plugin/configurable/AbstractEasyApiConfigurable.kt
@@ -4,6 +4,7 @@ import com.google.inject.Inject
 import com.intellij.openapi.options.SearchableConfigurable
 import com.intellij.openapi.project.Project
 import com.itangcent.common.kit.toJson
+import com.itangcent.common.logger.Log
 import com.itangcent.idea.plugin.api.export.core.EasyApiConfigReader
 import com.itangcent.idea.plugin.settings.SettingBinder
 import com.itangcent.intellij.config.ConfigReader
@@ -90,9 +91,9 @@ abstract class AbstractEasyApiConfigurable(private var myProject: Project?) : Se
         context?.stop()
     }
 
+    companion object : Log()
+
     init {
         LOG.info("create instance of ${this::class.qualifiedName}")
     }
 }
-
-private val LOG = com.intellij.openapi.diagnostic.Logger.getInstance(EasyApiConfigurable::class.java)

--- a/idea-plugin/src/main/kotlin/com/itangcent/idea/plugin/dialog/AbstractApiDashboardDialog.kt
+++ b/idea-plugin/src/main/kotlin/com/itangcent/idea/plugin/dialog/AbstractApiDashboardDialog.kt
@@ -7,6 +7,7 @@ import com.intellij.openapi.module.ModuleManager
 import com.intellij.openapi.project.Project
 import com.intellij.openapi.project.rootManager
 import com.intellij.psi.*
+import com.itangcent.common.logger.Log
 import com.itangcent.common.model.Doc
 import com.itangcent.common.model.MethodDoc
 import com.itangcent.common.model.Request
@@ -543,6 +544,7 @@ abstract class AbstractApiDashboardDialog : ContextDialog() {
                         psiResource.resource() as PsiMethod
                     )
                 }\n${(doc as Request).method}:${(doc as Request).path}"
+
                 else -> "${
                     PsiClassUtils.fullNameOfMethod(
                         psiResource.resourceClass()!!,
@@ -650,6 +652,6 @@ abstract class AbstractApiDashboardDialog : ContextDialog() {
             return super.createTransferable(component)
         }
     }
-}
 
-private val LOG = com.intellij.openapi.diagnostic.Logger.getInstance(AbstractApiDashboardDialog::class.java)
+    companion object : Log()
+}

--- a/idea-plugin/src/main/kotlin/com/itangcent/idea/plugin/dialog/ApiCallDialog.kt
+++ b/idea-plugin/src/main/kotlin/com/itangcent/idea/plugin/dialog/ApiCallDialog.kt
@@ -11,6 +11,7 @@ import com.intellij.ui.table.JBTable
 import com.intellij.util.ui.ComboBoxCellEditor
 import com.itangcent.cache.HttpContextCacheHelper
 import com.itangcent.common.constant.HttpMethod
+import com.itangcent.common.logger.Log
 import com.itangcent.common.logger.traceError
 import com.itangcent.common.model.FormParam
 import com.itangcent.common.model.Request
@@ -1083,6 +1084,7 @@ class ApiCallDialog : ContextDialog(), ApiCallUI {
                         val suffix = url.substring(dotIndex).substringBefore("\\//?&")
                         "$name.$suffix"
                     }
+
                     else -> url.substringAfterLast("/").substringBefore("?")
                 }
             }
@@ -1131,6 +1133,7 @@ class ApiCallDialog : ContextDialog(), ApiCallUI {
                     formatResult = formatResponse()
                     formatResult
                 }
+
                 else -> getRawResult()
             }
         }
@@ -1322,7 +1325,7 @@ class ApiCallDialog : ContextDialog(), ApiCallUI {
         return requestRawInfo
     }
 
-    companion object {
+    companion object : Log() {
         var CONTENT_TYPES: Array<String> = arrayOf(
             "",
             "application/json",
@@ -1351,5 +1354,3 @@ class ApiCallDialog : ContextDialog(), ApiCallUI {
         }
     }
 }
-
-private val LOG = com.intellij.openapi.diagnostic.Logger.getInstance(ApiCallDialog::class.java)

--- a/idea-plugin/src/main/kotlin/com/itangcent/idea/plugin/dialog/ApiDashboardDialog.kt
+++ b/idea-plugin/src/main/kotlin/com/itangcent/idea/plugin/dialog/ApiDashboardDialog.kt
@@ -2,10 +2,14 @@ package com.itangcent.idea.plugin.dialog
 
 import com.google.inject.Inject
 import com.intellij.openapi.ui.Messages
+import com.itangcent.common.logger.Log
 import com.itangcent.common.logger.traceError
 import com.itangcent.common.model.Doc
 import com.itangcent.common.model.Request
-import com.itangcent.common.utils.*
+import com.itangcent.common.utils.DateUtils
+import com.itangcent.common.utils.mapNotNull
+import com.itangcent.common.utils.mapToTypedArray
+import com.itangcent.common.utils.notNullOrBlank
 import com.itangcent.idea.icons.EasyIcons
 import com.itangcent.idea.icons.iconOnly
 import com.itangcent.idea.plugin.api.export.postman.PostmanCachedApiHelper
@@ -22,16 +26,15 @@ import com.itangcent.idea.utils.reload
 import com.itangcent.idea.utils.remove
 import com.itangcent.intellij.context.ActionContext
 import com.itangcent.intellij.context.Boundary
-import com.itangcent.intellij.extend.guice.PostConstruct
 import com.itangcent.intellij.extend.runWithContext
 import com.itangcent.intellij.extend.rx.eval
 import com.itangcent.intellij.extend.rx.from
-import org.apache.commons.lang3.exception.ExceptionUtils
 import java.awt.dnd.DnDConstants
 import java.awt.dnd.DropTarget
 import java.awt.dnd.DropTargetAdapter
 import java.awt.dnd.DropTargetDropEvent
-import java.awt.event.*
+import java.awt.event.MouseAdapter
+import java.awt.event.MouseEvent
 import java.util.*
 import java.util.concurrent.TimeUnit
 import javax.swing.*
@@ -433,10 +436,12 @@ class ApiDashboardDialog : AbstractApiDashboardDialog() {
             return
         }
         actionContext.runInSwingUI {
-            val newCollectionName = Messages.showInputDialog(this,
+            val newCollectionName = Messages.showInputDialog(
+                this,
                 "Input New Collection Name",
                 "New Collection",
-                Messages.getInformationIcon())
+                Messages.getInformationIcon()
+            )
             if (newCollectionName.isNullOrBlank()) return@runInSwingUI
 
             actionContext.runAsync {
@@ -477,10 +482,12 @@ class ApiDashboardDialog : AbstractApiDashboardDialog() {
         if (lastSelectedPathComponent != null) {
             val postmanNodeData = lastSelectedPathComponent.userObject ?: return
             actionContext.runInSwingUI {
-                val newCollectionName = Messages.showInputDialog(this,
+                val newCollectionName = Messages.showInputDialog(
+                    this,
                     "Input Sub Collection Name",
                     "Collection Name",
-                    Messages.getInformationIcon())
+                    Messages.getInformationIcon()
+                )
                 if (newCollectionName.isNullOrBlank()) return@runInSwingUI
 
                 actionContext.runAsync {
@@ -889,6 +896,6 @@ class ApiDashboardDialog : AbstractApiDashboardDialog() {
     override fun filterDoc(doc: Doc): Doc? {
         return doc.takeIf { it is Request }
     }
-}
 
-private val LOG = com.intellij.openapi.diagnostic.Logger.getInstance(ApiDashboardDialog::class.java)
+    companion object : Log()
+}

--- a/idea-plugin/src/main/kotlin/com/itangcent/idea/plugin/dialog/AskWithApplyAllDialog.kt
+++ b/idea-plugin/src/main/kotlin/com/itangcent/idea/plugin/dialog/AskWithApplyAllDialog.kt
@@ -1,5 +1,6 @@
 package com.itangcent.idea.plugin.dialog
 
+import com.itangcent.common.logger.Log
 import com.itangcent.idea.utils.SwingUtils
 import java.awt.EventQueue
 import java.awt.Window
@@ -80,10 +81,12 @@ class AskWithApplyAllDialog(owner: Window? = null) : JDialog(owner) {
         })
 
         // call onCancel() on ESCAPE
-        contentPane!!.registerKeyboardAction({ onCancel() },
+        contentPane!!.registerKeyboardAction(
+            { onCancel() },
             KeyStroke.getKeyStroke(KeyEvent.VK_ESCAPE, 0),
-            JComponent.WHEN_ANCESTOR_OF_FOCUSED_COMPONENT)
+            JComponent.WHEN_ANCESTOR_OF_FOCUSED_COMPONENT
+        )
     }
-}
 
-private val LOG = com.intellij.openapi.diagnostic.Logger.getInstance(AskWithApplyAllDialog::class.java)
+    companion object : Log()
+}

--- a/idea-plugin/src/main/kotlin/com/itangcent/idea/plugin/dialog/EasyApiSettingGUI.kt
+++ b/idea-plugin/src/main/kotlin/com/itangcent/idea/plugin/dialog/EasyApiSettingGUI.kt
@@ -7,6 +7,7 @@ import com.intellij.openapi.util.text.StringUtil
 import com.intellij.ui.MutableCollectionComboBoxModel
 import com.intellij.ui.table.JBTable
 import com.itangcent.cache.withoutCache
+import com.itangcent.common.logger.Log
 import com.itangcent.common.utils.*
 import com.itangcent.idea.icons.EasyIcons
 import com.itangcent.idea.icons.iconOnly
@@ -623,7 +624,7 @@ class EasyApiSettingGUI : AbstractEasyApiSettingGUI() {
         readPostmanCollections(settings)
     }
 
-    companion object {
+    companion object : Log() {
         const val basePath = ".easy_api"
 
         const val setting_path = "easy.api.setting.path"
@@ -635,5 +636,3 @@ class EasyApiSettingGUI : AbstractEasyApiSettingGUI() {
         private var DEFAULT_WORKSPACE = PostmanWorkspaceData(null, "")
     }
 }
-
-private val LOG = com.intellij.openapi.diagnostic.Logger.getInstance(YapiDashboardDialog::class.java)

--- a/idea-plugin/src/main/kotlin/com/itangcent/idea/plugin/dialog/YapiDashboardDialog.kt
+++ b/idea-plugin/src/main/kotlin/com/itangcent/idea/plugin/dialog/YapiDashboardDialog.kt
@@ -2,6 +2,7 @@ package com.itangcent.idea.plugin.dialog
 
 import com.google.inject.Inject
 import com.intellij.openapi.ui.Messages
+import com.itangcent.common.logger.Log
 import com.itangcent.common.model.Doc
 import com.itangcent.common.utils.notNullOrBlank
 import com.itangcent.idea.icons.EasyIcons
@@ -15,22 +16,18 @@ import com.itangcent.idea.swing.EasyApiTreeCellRenderer
 import com.itangcent.idea.swing.IconCustomized
 import com.itangcent.idea.swing.ToolTipAble
 import com.itangcent.idea.utils.SwingUtils
-import com.itangcent.idea.utils.initAfterShown
 import com.itangcent.idea.utils.isDoubleClick
 import com.itangcent.idea.utils.reload
-import com.itangcent.intellij.constant.EventKey
 import com.itangcent.intellij.context.ActionContext
 import com.itangcent.intellij.extend.*
-import com.itangcent.intellij.extend.guice.PostConstruct
 import com.itangcent.intellij.extend.rx.from
 import org.apache.commons.lang3.exception.ExceptionUtils
 import java.awt.dnd.DnDConstants
 import java.awt.dnd.DropTarget
 import java.awt.dnd.DropTargetAdapter
 import java.awt.dnd.DropTargetDropEvent
-import java.awt.event.*
-import java.util.*
-import java.util.concurrent.Future
+import java.awt.event.MouseAdapter
+import java.awt.event.MouseEvent
 import java.util.concurrent.TimeUnit
 import javax.swing.*
 import javax.swing.tree.DefaultMutableTreeNode
@@ -505,12 +502,14 @@ class YapiDashboardDialog : AbstractApiDashboardDialog() {
                 is YapiApiNodeData -> {
                     syncYapiNode(yapiNodeData.getParentNodeData())
                 }
+
                 is YapiProjectNodeData -> {
                     //clear
                     yapiNodeData.removeAllSub()
                     //reload
                     loadYapiProject(yapiNodeData)
                 }
+
                 is YapiCartNodeData -> {
                     //clear
                     yapiNodeData.removeAllSub()
@@ -751,6 +750,7 @@ class YapiDashboardDialog : AbstractApiDashboardDialog() {
                                         Thread.sleep(200)
                                         syncYapiNode(targetNodeData.getRootNodeData())
                                     }
+
                                     Messages.NO -> {
                                         export(fromProjectData, targetNodeData, cartId)
                                         Thread.sleep(200)
@@ -793,6 +793,5 @@ class YapiDashboardDialog : AbstractApiDashboardDialog() {
     }
 
     //endregion handle drop--------------------------------------------------------
+    companion object : Log()
 }
-
-private val LOG = com.intellij.openapi.diagnostic.Logger.getInstance(YapiDashboardDialog::class.java)

--- a/idea-plugin/src/main/kotlin/com/itangcent/idea/plugin/render/AdaptiveMarkdownRender.kt
+++ b/idea-plugin/src/main/kotlin/com/itangcent/idea/plugin/render/AdaptiveMarkdownRender.kt
@@ -2,6 +2,7 @@ package com.itangcent.idea.plugin.render
 
 import com.google.inject.Inject
 import com.google.inject.Singleton
+import com.itangcent.common.logger.Log
 import com.itangcent.common.logger.traceError
 import com.itangcent.intellij.config.ConfigReader
 import com.itangcent.intellij.context.ActionContext
@@ -85,6 +86,6 @@ class AdaptiveMarkdownRender : MarkdownRender {
         }
         return null
     }
-}
 
-private val LOG = com.intellij.openapi.diagnostic.Logger.getInstance(AdaptiveMarkdownRender::class.java)
+    companion object : Log()
+}

--- a/idea-plugin/src/main/kotlin/com/itangcent/idea/sqlite/SqliteDataResourceHelper.kt
+++ b/idea-plugin/src/main/kotlin/com/itangcent/idea/sqlite/SqliteDataResourceHelper.kt
@@ -99,8 +99,10 @@ class SqliteDataResourceHelper {
         override fun get(name: ByteArray): ByteArray? {
             return try {
                 return sqLiteDataSourceHandle.read {
-                    it.execute<String?>("SELECT * FROM $cacheName WHERE HASH = '${name.contentHashCode()}'" +
-                            " AND NAME = '${name.encodeBase64()}' LIMIT 1") { resultSet -> resultSet.getString("VALUE") }
+                    it.execute<String?>(
+                        "SELECT * FROM $cacheName WHERE HASH = '${name.contentHashCode()}'" +
+                                " AND NAME = '${name.encodeBase64()}' LIMIT 1"
+                    ) { resultSet -> resultSet.getString("VALUE") }
                         ?.decodeBase64()
                 }
             } catch (e: Exception) {
@@ -170,8 +172,10 @@ class SqliteDataResourceHelper {
                 val hash = name.contentHashCode()
                 var expired: Long? = null
                 val value = sqLiteDataSourceHandle.read { sqLiteDataSource ->
-                    sqLiteDataSource.execute<String?>("SELECT * FROM $cacheName WHERE HASH = '$hash'" +
-                            " AND NAME = '$base64Name' LIMIT 1") { resultSet ->
+                    sqLiteDataSource.execute<String?>(
+                        "SELECT * FROM $cacheName WHERE HASH = '$hash'" +
+                                " AND NAME = '$base64Name' LIMIT 1"
+                    ) { resultSet ->
                         val expiredInResult = resultSet.getLong("EXPIRED")
                         return@execute if (notExpired(expiredInResult)) {
                             resultSet.getString("VALUE")
@@ -184,10 +188,11 @@ class SqliteDataResourceHelper {
                 if (expired != null) {//delete expired row
                     try {
                         sqLiteDataSourceHandle.write {
-                            it.execute("DELETE FROM $cacheName WHERE " +
-                                    "HASH = $hash " +
-                                    "AND NAME = '$base64Name'" +
-                                    "AND EXPIRED = '$expired'"
+                            it.execute(
+                                "DELETE FROM $cacheName WHERE " +
+                                        "HASH = $hash " +
+                                        "AND NAME = '$base64Name'" +
+                                        "AND EXPIRED = '$expired'"
                             ) {}
                         }
                     } catch (e: Exception) {
@@ -331,5 +336,3 @@ fun SqliteDataResourceHelper.ExpiredBeanDAO.set(key: String, value: String, expi
 fun SqliteDataResourceHelper.ExpiredBeanDAO.delete(str: String) {
     this.delete(str.toByteArray(Charsets.UTF_8))
 }
-
-private val LOG = com.intellij.openapi.diagnostic.Logger.getInstance(SqliteDataResourceHelper::class.java)

--- a/idea-plugin/src/main/kotlin/com/itangcent/idea/utils/CustomizedPsiClassHelper.kt
+++ b/idea-plugin/src/main/kotlin/com/itangcent/idea/utils/CustomizedPsiClassHelper.kt
@@ -5,6 +5,8 @@ import com.intellij.psi.PsiClass
 import com.intellij.psi.PsiElement
 import com.intellij.psi.PsiField
 import com.itangcent.common.constant.Attrs
+import com.itangcent.common.logger.Log
+import com.itangcent.common.logger.traceWarn
 import com.itangcent.common.utils.*
 import com.itangcent.idea.plugin.api.export.AdditionalField
 import com.itangcent.idea.plugin.api.export.core.ClassExportRuleKeys
@@ -74,7 +76,7 @@ open class CustomizedPsiClassHelper : ContextualPsiClassHelper() {
         val obj = try {
             GsonExUtils.fromJson<Any>(valueText)
         } catch (e: Exception) {
-            LOG.warn("failed parse json:\n$valueText\n", e)
+            LOG.traceWarn("failed parse json:\n$valueText\n", e)
             return
         }
         if (obj.isOriginal()) {
@@ -113,7 +115,6 @@ open class CustomizedPsiClassHelper : ContextualPsiClassHelper() {
         }
         return super.ignoreField(psiField)
     }
+
+    companion object : Log()
 }
-
-
-private val LOG = com.intellij.openapi.diagnostic.Logger.getInstance(CustomizedPsiClassHelper::class.java)

--- a/idea-plugin/src/main/kotlin/com/itangcent/idea/utils/JacksonUtils.kt
+++ b/idea-plugin/src/main/kotlin/com/itangcent/idea/utils/JacksonUtils.kt
@@ -5,10 +5,11 @@ import com.fasterxml.jackson.annotation.PropertyAccessor
 import com.fasterxml.jackson.databind.DeserializationFeature
 import com.fasterxml.jackson.databind.ObjectMapper
 import com.fasterxml.jackson.databind.SerializationFeature
-import com.fasterxml.jackson.databind.jsontype.impl.LaissezFaireSubTypeValidator
+import com.itangcent.common.logger.Log
+import com.itangcent.common.logger.traceError
 
 
-object JacksonUtils {
+object JacksonUtils : Log() {
 
     private val objectMapper: ObjectMapper = ObjectMapper()
 
@@ -42,10 +43,8 @@ object JacksonUtils {
                 Class.forName(json.substring(0, split)) as Class<T>
             )
         } catch (e: Exception) {
-            LOG.error("failed parse json: [$json]", e)
+            LOG.traceError("failed parse json: [$json]", e)
             null
         }
     }
 }
-
-private val LOG = com.intellij.openapi.diagnostic.Logger.getInstance(JacksonUtils::class.java)

--- a/idea-plugin/src/main/kotlin/com/itangcent/spi/SpiCompositeLoader.kt
+++ b/idea-plugin/src/main/kotlin/com/itangcent/spi/SpiCompositeLoader.kt
@@ -14,7 +14,6 @@ import kotlin.reflect.full.findAnnotation
 
 object SpiCompositeLoader {
 
-    @Suppress("UNCHECKED_CAST")
     inline fun <reified S : Any> loadComposite(): S {
         val cls = S::class
         val loadServices = load<S>(ActionContext.getContext()!!)

--- a/idea-plugin/src/test/kotlin/com/itangcent/test/ResultLoader.kt
+++ b/idea-plugin/src/test/kotlin/com/itangcent/test/ResultLoader.kt
@@ -1,14 +1,12 @@
 package com.itangcent.test
 
-import com.itangcent.common.logger.ILogger
-import com.itangcent.common.logger.ILoggerProvider
-import com.itangcent.common.spi.SpiUtils
+import com.itangcent.common.logger.Log
 import com.itangcent.common.utils.ResourceUtils
 import com.itangcent.mock.toUnixString
 import com.itangcent.test.TimeZoneKit.fixTimeZone
 import kotlin.reflect.KClass
 
-object ResultLoader {
+object ResultLoader : Log() {
 
     fun load(): String {
         val grandCallerClass = com.intellij.util.ReflectionUtil.getGrandCallerClass()!!
@@ -17,7 +15,7 @@ object ResultLoader {
             .replace('$', '.')
         val resource = ResourceUtils.readResource("result/$rawName.txt")
         if (resource.isEmpty()) {
-            LOG?.warn("no resource [result/$rawName.txt] be found")
+            LOG.warn("no resource [result/$rawName.txt] be found")
         }
         return resource.toUnixString().fixTimeZone()
     }
@@ -38,11 +36,8 @@ object ResultLoader {
         val fileName = "$rawName.$name"
         val resource = ResourceUtils.readResource("result/$fileName.txt")
         if (resource.isEmpty()) {
-            LOG?.warn("no resource [result/$fileName.txt] be found")
+            LOG.warn("no resource [result/$fileName.txt] be found")
         }
         return resource.toUnixString().fixTimeZone()
     }
 }
-
-
-private val LOG: ILogger? = SpiUtils.loadService(ILoggerProvider::class)?.getLogger(ResultLoader::class)

